core/Pos/Binary/Core/Address.hs

```diff
@@ -1,95 +1,189 @@
+import           Pos.Core.Types        (AddrAttributes (..), AddrSpendingData (..),
+                                        AddrStakeDistribution (..), AddrType (..),
+                                        Address (..), Address' (..))
+import           Pos.Data.Attributes   (Attributes (..), decodeAttributes,
+                                        encodeAttributes)
+
+----------------------------------------------------------------------------
+-- Helper types serialization
+----------------------------------------------------------------------------
+
+-- Helper function to avoid writing `:: Word8`.
+w8 :: Word8 -> Word8
+w8 = identity
+{-# INLINE w8 #-}
+
+instance Bi AddrType where
+    encode =
+        encode @Word8 . \case
+            ATPubKey -> 0
+            ATScript -> 1
+            ATRedeem -> 2
+            ATUnknown tag -> tag
+    decode =
+        decode @Word8 <&> \case
+            0 -> ATPubKey
+            1 -> ATScript
+            2 -> ATRedeem
+            tag -> ATUnknown tag
+
+{- NOTE: Address spending data serialization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-import           Codec.CBOR.Encoding (Encoding)
-import qualified Codec.CBOR.Write    as CBOR.Write
-import           Data.Default        (def)
-import           Data.Digest.CRC32   (CRC32 (..))
-import           Data.Word           (Word8)
-import           Formatting          (format, shown, (%))
-import           Pos.Binary.Class    (Bi (..), deserialize', encodeListLen, enforceSize,
-                                      serialize')
-import           Pos.Binary.Crypto   ()
-import           Pos.Core.Types      (AddrPkAttrs (..), Address (..))
-import           Pos.Data.Attributes (Attributes, decodeAttributes, encodeAttributes)
+An address is serialized as a tuple consisting of:
 
-{- NOTE: Address serialization
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+1. One-byte tag.
+2. Data dependent on tag.
 
-An address is serialized as follows:
+If tag is 0, 1 or 2, the type of spending data is 'PubKeyASD',
+'ScriptASD' or 'RedeemASD' respectively.
 
-    <one-byte tag><length of content><content>
+If tag is greater than 2, the data is decoded as a plain 'ByteString'.
 
-This lets us have backwards compatibility. For instance, if a newer version
-of CSL adds an address with tag 5:
+This lets us have backwards compatibility. For instance, if a newer
+version of CSL adds a new type of spending data with tag 3, then older
+versions would deserialize it as follows:
 
-    data Address
-        = ...
-        | SuperAddress A B C
+    UnknownASD 3 <some bytes>
+-}
 
-then older versions would deserialize it as follows:
+instance Bi AddrSpendingData where
+    encode =
+        \case
+            PubKeyASD pk -> encode (w8 0, pk)
+            ScriptASD script -> encode (w8 1, script)
+            RedeemASD redeemPK -> encode (w8 2, redeemPK)
+            UnknownASD tag payload -> encode (tag, payload)
+    decode = do
+        enforceSize "AddrSpendingData" 2
```

Comment:
I am not familiar with CBOR. does it work for tuples as for `encode field1 <> encode field2 <> ...`?